### ISSUE TITLE
[BUG] hang when node is crash or killed

### ISSUE THREAD

**Issue by NiuBlibing** [microsoft/playwright-python#1779]
**Context:**
- Playwright Version: playwright==1.30.0
- Operating System: Linux (Linux manjaro 5.15.94-1-MANJARO)
- Python version: 3.10.9
- Browser: Chromium

**Code Snippet**

```python
import asyncio

from playwright.async_api import async_playwright, PlaywrightContextManager

from pympler import muppy
from pympler import summary
import gc
import tracemalloc
import logging
logger = logging.getLogger(__name__)


async def handle(route, request):
    headers = {
    }
    await route.continue_(headers=headers)
def _make_request_logger():
    async def _log_request(request) -> None:
        referrer = await request.header_value("referer")
        del referrer

    return _log_request
def close_browser_call() -> None:
    logger.debug("Browser closed")

async def main():
    tracemalloc.start()

    async with async_playwright() as p:
        browser = await p.chromium.launch()
        browser.on('disconnected', close_browser_call)
        context = await browser.new_context(ignore_https_errors=True)
        snapshot1 = tracemalloc.take_snapshot()

        for i in range(100):
            page = await context.new_page()
            await page.route(
            "**",
            handle)
            page.on("request",_make_request_logger())
            try:
                await page.goto("https://httpstat.us/200?sleep=3", wait_until="networkidle", timeout=100)
            except Exception as e:
                print(e)
            finally:
                await page.close()
                del page
            await context.close()
            del context
            context = await browser.new_context(ignore_https_errors=True)
            gc.collect()
            snapshot2 = tracemalloc.take_snapshot()
            top_stats = snapshot2.compare_to(snapshot1, 'lineno')
            print("[ Top 10 differences ]")
            for stat in top_stats[:10]:
                print(stat)
            del snapshot2
            
            

asyncio.run(main())

```

**Describe the bug**

 During mass crawling, the node may crash unexpectedly or be terminated by the oom-killer, and the program may get stuck.
I use `kill -9` command to stop the node, then the following shows and the program hangs:
```log
/home/test/source/shanhai/maliciousDownloader/malicious_downloader/venv/lib/python3.10/site-packages/playwright/driver/playwright.sh: line 6: 21110 Killed                  "$PLAYWRIGHT_NODEJS_PATH" "$SCRIPT_PATH/package/lib/cli/cli.js" "$@"
```

**Comment by mxschmitt** [comment]
The context which you create at the bottom via `context = await browser.new_context(ignore_https_errors=True)` you are never closing, thats most likely the issue.

**Comment by NiuBlibing** [comment]
> The context which you create at the bottom via you are never closing, thats most likely the issue.`context = await browser.new_context(ignore_https_errors=True)`

The context is created and will be deleted in next loop. The kill command is sent during the loop.

**Comment by mxschmitt** [comment]
Can you provide us a repro how node naturally crashes or gets killed? If its because you overbook your CPU/memory and then node ends up in a OOM, we unlikely can help there, since this is an overbooking issue on your end.

**Comment by NiuBlibing** [comment]
I'm using [scrapy-playwright](https://github.com/scrapy-plugins/scrapy-playwright) to do broad crawling.Because of the memory leak such as  [bug1](https://github.com/microsoft/playwright/issues/6319), the oom is unavoidable and the node process may be killed first by system kernel then everything hangs.

**Comment by petroslamb** [comment]
This continues to be an issue for me using `sync_playwright`.  The fix did not completely mend the problem, as in some cases the OOM kill signal to the node/chrome process is the last thing that happens.  

By that I mean that the exception `Connection closed while reading from the driver`, is not always thrown. This leads to a zombie container, with a python process left, while node/chrome are killed. My app lives in a container, in a pod, in a k8s cluster btw.

I am guessing here, but it could be that the runtime is in a different part of the code, when the OOM kill is received, so the exception is never raised? Also (unrelated), when thrown, a generic expression in thrown, not one derived by `playwright.Error`.

I know about the open/close context issue and the memory leak. Still the OOM can happen regardless and I am not sure that dumb-init or the --init flag can fix it, as the OOM kill does not go to the PID 1, which is python.

Any suggestions for workarounds?

**Comment by milan-cp-dev** [comment]
Hey @NiuBlibing since I am using scrapy-playwright as well, I have handled it by using their memory usage extension. Such extension follows child processes to calculate memory usage. Their extension will raise Error that scrapy won't catch until you convert it to some scrapy known exception. That should establish a stable scrape. I have opened an issue on their repo outlining setup. Maybe they will patch it as well to have it in their library. https://github.com/scrapy-plugins/scrapy-playwright/issues/331
